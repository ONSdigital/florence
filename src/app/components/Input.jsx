import React, { Component } from "react";
import PropTypes from "prop-types";
import clsx from "clsx";

const propTypes = {
    id: PropTypes.string.isRequired,
    name: PropTypes.string,
    label: PropTypes.string,
    type: PropTypes.string,
    onChange: PropTypes.func,
    onClearValue: PropTypes.func,
    onBlur: PropTypes.func,
    error: PropTypes.string,
    fileError: PropTypes.string,
    displayInputAsErrored: PropTypes.bool,
    disabled: PropTypes.bool,
    isFocused: PropTypes.bool,
    inline: PropTypes.bool,
    accept: PropTypes.string,
    value: PropTypes.string,
    helpMessage: PropTypes.string,
    min: PropTypes.string,
    max: PropTypes.string,
    allowAutoComplete: PropTypes.bool,
    placeholder: PropTypes.string,
    disableShowPasswordText: PropTypes.bool,
    reverseLabelOrder: PropTypes.bool,
};

const defaultProps = {
    type: "text",
    disabled: false,
    isFocused: false,
};

export default class Input extends Component {
    constructor(props) {
        super(props);

        this.state = {
            type: this.props.type,
            displayShowHide: (!this.props.disableShowPasswordText && this.props.type) === "password",
        };
    }

    showHide = e => {
        e.preventDefault();
        e.stopPropagation();
        this.setState({
            type: this.state.type === "text" ? "password" : "text",
        });
    };

    moveCaretToEnd = event => {
        // Move caret to the end of the value on input focus
        const val = event.target.value;
        event.target.value = "";
        event.target.value = val;
    };

    getFormClasses() {
        let formClasses = "form__input";
        if (this.props.error || this.props.displayInputAsErrored) {
            formClasses += " form__input--error";
        }
        if (this.props.inline) {
            formClasses += " form__input--flush";
        }
        if (this.props.type === "checkbox") {
            formClasses += " checkbox";
        }
        return formClasses;
    }

    getInputClasses() {
        let inputClasses = "input";
        if (this.state.displayShowHide) {
            inputClasses += " input--show-hide";
        }
        if (this.props.type === "checkbox") {
            inputClasses += " checkbox__input visually-hidden";
        }
        return inputClasses;
    }

    renderInput() {
        switch (this.props.type) {
            case "textarea":
                return (
                    <textarea
                        id={this.props.id}
                        className="input input__textarea"
                        name={this.props.name || this.props.id}
                        disabled={this.props.disabled}
                        onChange={this.props.onChange}
                        onBlur={this.props.onBlur}
                        autoFocus={this.props.isFocused}
                        placeholder={this.props.inline && this.props.label}
                        value={this.props.value}
                    />
                );
            case "clear":
                return (
                    <span className="deleteIcon">
                        <input
                            id={this.props.id}
                            type="text"
                            className={this.getInputClasses()}
                            name={this.props.name || this.props.id}
                            disabled={this.props.disabled}
                            onChange={this.props.onChange}
                            onBlur={this.props.onBlur}
                            autoFocus={this.props.isFocused}
                            onFocus={this.moveCaretToEnd}
                            placeholder={this.props.inline ? this.props.label : this.props.placeholder}
                            accept={this.props.accept}
                            value={this.props.value}
                            min={this.props.min}
                            max={this.props.max}
                            autoComplete={!this.props.allowAutoComplete && "new-password"}
                        />
                        <span onClick={this.props.onClearValue}>x</span>
                    </span>
                );
            default:
                return (
                    <input
                        id={this.props.id}
                        /* If using old 'Show' within input then use 'state' as it is managed within the input component
                         * otherwise use the prop which is managed by the parent */
                        type={this.state.displayShowHide ? this.state.type : this.props.type}
                        className={this.getInputClasses()}
                        name={this.props.name || this.props.id}
                        disabled={this.props.disabled}
                        onChange={this.props.onChange}
                        onBlur={this.props.onBlur}
                        autoFocus={this.props.isFocused}
                        onFocus={this.moveCaretToEnd}
                        placeholder={this.props.inline ? this.props.label : this.props.placeholder}
                        accept={this.props.accept}
                        value={this.props.value}
                        min={this.props.min}
                        max={this.props.max}
                        // setting autocomplete to "new-password" will/should tell Google this is a
                        //  "sign up form" type of input and not use the auto complete, more info:
                        // https://www.chromium.org/developers/design-documents/form-styles-that-chromium-understands
                        autoComplete={!this.props.allowAutoComplete && "new-password"}
                    />
                );
        }
    }

    renderLabel() {
        return (
            <label className={clsx("form__label", { "checkbox__label font-weight--300": this.props.type === "checkbox" })} htmlFor={this.props.id}>
                {this.props.label}
            </label>
        );
    }

    render() {
        const formClasses = this.getFormClasses();
        return (
            <div className={formClasses}>
                {!this.props.inline && !this.props.reverseLabelOrder && this.renderLabel()}
                {this.props.error && (
                    <div id={`input-error-${this.props.id}`} data-testid={this.props.id} className="error-msg" role="alert">
                        {this.props.error}
                    </div>
                )}
                {this.props.helpMessage && <div className="help-msg">{this.props.helpMessage}</div>}
                {this.props.fileError && (
                    <div id={`input-error-${this.props.id}`} data-testid={this.props.id} className="error-msg" role="alert">
                        {this.props.fileError}
                    </div>
                )}
                {this.renderInput()}
                {this.state.displayShowHide && (
                    <span className="btn btn--password" onClick={this.showHide} onKeyPress={this.showHide} tabIndex="0" role="button">
                        {this.state.type === "text" ? "Hide" : "Show"}
                    </span>
                )}
                {this.props.inline && this.props.reverseLabelOrder && this.renderLabel()}
            </div>
        );
    }
}

Input.propTypes = propTypes;
Input.defaultProps = defaultProps;
