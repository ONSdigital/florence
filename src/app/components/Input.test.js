import React from "react";
import Input from "./Input.jsx";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

test("Input without type renders a text input", () => {
    const props = {
        id: "test-input",
        label: "Test input",
    };
    const component = renderer.create(<Input id={props.id} label={props.label} />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("Password prop creates a password input", () => {
    const props = {
        id: "test-password-input",
        label: "Password",
        type: "password",
    };
    const component = renderer.create(<Input {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("Password show/hide button text toggles on click", () => {
    const props = {
        id: "test-password-input",
        label: "Password",
        type: "password",
    };
    const component = shallow(<Input {...props} />);
    const buttonText = component.find(".btn--password").text();
    component.find(".btn--password").simulate("click", {
        stopPropagation: () => undefined,
        preventDefault: () => {},
    });
    expect(component.find(".btn--password").text()).not.toEqual(buttonText);
});

test('Password show/hide button toggles input type state between "password" and "text"', () => {
    const props = {
        id: "test-password-input",
        label: "Password",
        type: "password",
    };
    const component = shallow(<Input {...props} />);
    expect(component.state("type")).toEqual("password");
    component.find(".btn--password").simulate("click", {
        stopPropagation: () => undefined,
        preventDefault: () => {},
    });
    expect(component.state("type")).toEqual("text");
});
