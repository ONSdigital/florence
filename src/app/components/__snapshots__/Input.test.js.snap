// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input without type renders a text input 1`] = `
<div
  className="form__input"
>
  <label
    className="form__label"
    htmlFor="test-input"
  >
    Test input
    : 
  </label>
  
  <input
    autoFocus={false}
    className="input input__text"
    disabled={false}
    id="test-input"
    name="test-input"
    onChange={undefined}
    placeholder=""
    type="text"
  />
  
</div>
`;

exports[`Password prop creates a password input 1`] = `
<div
  className="form__input"
>
  <label
    className="form__label"
    htmlFor="test-password-input"
  >
    Password
    : 
  </label>
  
  <input
    autoFocus={false}
    className="input input__text"
    disabled={false}
    id="test-password-input"
    name="test-password-input"
    onChange={undefined}
    placeholder=""
    type="password"
  />
  <span
    className="btn btn--password"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    Show
  </span>
</div>
`;
